test/unified/runner.js

```diff
@@ -0,0 +1,366 @@
+    const results = [];
+    let foundError = undefined;
+    return tasks.reverse().reduce((cb, task) => {
+      return () => {
+        if (foundError) return cb(foundError);
+        task((err, result) => {
+          if (err) {
+            foundError = err;
+            return cb(err);
+          }
+          results.push(result);
+          return cb(undefined, results);
+        });
+      };
+    }, cb)();
+  }
+  /** @see https://gist.github.com/bisubus/2da8af7e801ffd813fab7ac221aa7afc#file-pick-es2015-js */
+  static pick(obj, picked) {
+    return Object.keys(obj)
+      .filter(key => picked.indexOf(key) >= 0)
+      .reduce((newObj, key) => Object.assign(newObj, { [key]: obj[key] }), {});
+  }
+  resetEntityMap(cb) {
+    this.entities = [];
+    this.clientEvents = [];
+    this.clientEntityIds = [];
+    return cb();
+  }
+  closeClients(cb) {
+    const cbs = this.clientEntityIds.map(id => {
+      return cb => {
+        const client = this.getEntity(id);
+        return client.close(true, cb);
+      };
+    });
+    return Runner.callbacksAll(cbs, cb);
+  }
+  static singleOrArray(value) {
+    if (!value) return [];
+    if (Array.isArray(value)) return value;
+    return [value];
+  }
+  static createMetadata(req) {
+    const topology = Runner.singleOrArray(req.topology);
+    const minServerVersion = req.minServerVersion;
+    const maxServerVersion = req.maxServerVersion;
+    let mongodb = [];
+    if (minServerVersion) mongodb.push(`>=${Version.clean(minServerVersion)}`);
+    if (maxServerVersion) mongodb.push(`<=${Version.clean(maxServerVersion)}`);
+    const requires = {};
+    if (topology.length) requires.topology = topology;
+    if (mongodb.length) requires.mongodb = mongodb.join(' ');
+    return { requires };
+  }
+  static createTestDescription(test, data) {
+    let description = [test.description];
+    if (data.topology && data.topology.length) description.push(data.topology.join(','));
+    if (data.mongodb) description.push(data.mongodb);
+    return description.join(' - ');
+  }
+  static handleSpec(spec) {
+    const r = new Runner(spec);
+    describe(r.spec.description, () => {
+      r.spec.tests.forEach(test => {
+        r.spec.runOnRequirements.forEach(req => {
+          const metadata = Runner.createMetadata(req);
+          const description = Runner.createTestDescription(test, metadata.requires);
+          it(description, metadata, function (done) {
+            Runner.callbacksAll(
+              [
+                cb => r.createTopClient(cb),
+                cb => r.createEntities(cb),
+                cb => r.removeInitialData(cb),
+                cb => r.createInitialData(cb),
+                cb => r.handleTest(test, cb),
+                cb => r.closeClients(cb),
+                cb => r.resetEntityMap(cb)
+              ],
+              done
+            );
+          });
+        });
+      });
+    });
+  }
+}
+
+module.exports = {
+  Version,
+  Runner
```

Comment:
Another desirable quality (IMO) of the existing spec runner is it's ability to fit seamlessly into our existing mocha tests. You can mix the runner into a test file which also includes prose tests, and it looks pretty natural:
```
describe('Transactions', function () {
  describe('spec tests', function () {
    const testContext = new TransactionsRunnerContext();      
    const testSuites = loadSpecTests(suiteSpec.specPath);
    after(() => testContext.teardown());
    before(function () {
      return testContext.setup(this.configuration);
    });

    generateTopologyTests(testSuites, testContext, (spec, config) => {
        // NODE-2574: remove this when HELP-15010 is resolved
        if (config.topologyType === 'Sharded' && semver.satisfies(config.version, '>=4.4')) {
          return false;
        }

        return true;
    });
  });

  it('should do something else', function(done) { ... });
  it('should do something else', function(done) { ... });
  it('should do something else', function(done) { ... });
  it('should do something else', function(done) { ... });
});
```

Is there a goal in this rewrite to preserve any of that flexibility, or is it your opinion that we don't benefit from these features?

Reply:
All of this functionality is encapsulated in `Runner.handleSpec(example)` which can be run directly inside mocha.

If you choose to be declarative in the mocha file as you have, you could run the same code I run in that method.

```js
const r = new Runner(spec);
describe(r.spec.description, () => {
  r.spec.tests.forEach(test => {
    r.spec.runOnRequirements.forEach(req => {
      const metadata = Runner.createMetadata(req);
      const description = Runner.createTestDescription(test, metadata.requires);
      it(description, metadata, function (done) {
        Runner.callbacksAll(
          [
            cb => r.createTopClient(cb),
            cb => r.createEntities(cb),
            cb => r.removeInitialData(cb),
            cb => r.createInitialData(cb),
            cb => r.handleTest(test, cb),
            cb => r.closeClients(cb),
            cb => r.resetEntityMap(cb)
          ],
          done
        );
      });
    });
  });
});
```